### PR TITLE
support deprecated attribute in type constructors

### PR DESCRIPTION
**Pull request by Wilbert-mad**
Addressing: #3698

* [X] Analysis should report error when a custom types' variants are all deprecated
* [X] Parser should error if an unsupported attribute is used on a variant
* [x] LSP warning for deprecated variants
* [x] Formator allow attributes for type constructors

### CODE REVIEW

File: compiler-core/src/parse.rs
Comment (Wilbert-mad) [inline compiler-core/src/parse.rs:2209]: Wondering here whether we include the whole type or just up to the end of the type name?

```
  ┌─ /src/parse/error.gleam:2:1
  │  
2 │ ╭ type Wibble {
3 │ │     @deprecated("1")
4 │ │     Wibble1
5 │ │     @deprecated("2")
6 │ │     Wibble2
7 │ │ }
  │ ╰─^ Can't deprecate all variants of a type.
```


```
  ┌─ /src/parse/error.gleam:2:1
  │  
2 │ type Wibble {
  │ ^^^^^^^^^^^ Can't deprecate all variants of a type.
```

Comment (Wilbert-mad) [reply compiler-core/src/parse.rs:2209]: decided to emit the latter--single-line error. Considering if a custom type might be too and redundant to print.

File: compiler-core/src/parse.rs
Comment (giacomocavalieri) [inline compiler-core/src/parse.rs:2208]: I think this error should happen during the analysis phase and not in the parser, if an error happens during parsing it would stop the compiler from being able to analyse the rest of the code

File: compiler-core/src/parse/error.rs
Comment (giacomocavalieri) [inline compiler-core/src/parse/error.rs:314]: I think I'd rather have an error that explicitly says that the attribute cannot be used on a variant for each of the attributes rather than a single error for all of those.

```
pub type Wibble {
  @external(javascript, "", "")
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ This attribute cannot be used on a type variant
  Wobble
}
```

Also I would not include that hint, it doesn't make much sense to suggest using `deprecated` if I wrote an `@external` or `@internal` attribute

File: compiler-core/src/parse/snapshots/gleam_core__parse__tests__deprecation_attribute_on_all_type_variants.snap
Comment (giacomocavalieri) [inline compiler-core/src/parse/snapshots/gleam_core__parse__tests__deprecation_attribute_on_all_type_variants.snap:17]: Really nice suggestion! Usually in the error reporting the convention is that multiline code snippets are indented by 4 spaces and preceded by an empty line, so this should be:
```suggestion
Consider deprecating the type as a whole:

    @deprecated("message")
    type Wibble {
      Wobble1
      Wobble2
    }
```

**Review by giacomocavalieri — COMMENTED** [review]
This is looking nice! I left a couple of comments inline

File: compiler-core/src/parse.rs
Comment (lpil) [reply compiler-core/src/parse.rs:2208]: Yup!

**Comment by lpil** [comment]
When you're ready for review please un-draft this PR 💜

File: compiler-core/src/format.rs
Comment (Wilbert-mad) [inline compiler-core/src/format.rs:1589]: I'm having issues getting the formatting to work here.
For example, the following is formatting to this 
```gleam
type Fooing {
  @deprecated("")
  Bar(//test
  a: Int, b: Int)
  Foo
}
```
instead of:
```gleam
type Fooing {
  @deprecated("")
  Bar(
    //test
    a: Int,
    b: Int,
  )
  Foo
}
```

**Comment by lpil** [comment]
Is this ready for a review? :)

**Comment by Wilbert-mad** [comment]
Yes! :D

File: compiler-core/src/error.rs
Comment (GearsDatapacks) [inline compiler-core/src/error.rs:3100]:
```suggestion
                                title: "Deprecating all variants of a type is not allowed.".into(),
```

File: CHANGELOG.md
Comment (GearsDatapacks) [inline CHANGELOG.md:22]:
```suggestion
- The compiler now allows deprecating specific constructors of a custom type using the `@deprecated` attribute:
```

**Comment by GearsDatapacks** [comment]
Just so you know, merge commits are not allowed in this repository. You'll need to rebase before Louis will accept it

**Comment by Wilbert-mad** [comment]
Alright, that should be good. Sorry about that. I'll leave it up to y'all from here.

**Comment by giacomocavalieri** [comment]
One thing is missing is version tracking. If someone is using the deprecated attribute on a constructor we want to make sure the version in their gleam.toml is `>= 1.6` (if this gets merged before the 1.6 release).

This would require adding a new variant `DeprecatedAttributeOnConstructor` to the `FeatureKind` enum (<https://github.com/gleam-lang/gleam/blob/9ee71a11393de9ebb1eb4a43dbebdea2def182f0/compiler-core/src/type_/error.rs#L834-L842>) and track its usage during the analysis phase I think

**Comment by Wilbert-mad** [comment]
image
In the analysis would it be best to warn on the `@deprecated` directly which would require the `RecordConstructor` to carry around a `SrcSpan` under `deprecation_loc: Option<SrcSpan>` or as shown in the top image using the `RecordConstructor`'s `location` `SrcSpan`. The lower image is the former. Reasonably carrying a SrcSpan wouldn't be as large data, but it would only serve its purpose once. (I can do either though just want feed back ig)
image

**Comment by Wilbert-mad** [comment]
Done!

File: CHANGELOG.md
Comment (lpil) [inline CHANGELOG.md:22]: Let's call them variants

File: CHANGELOG.md
Comment (lpil) [inline CHANGELOG.md:29]: Can we give these some other names? `I` is a bit hard to read. Perhaps we can use the name of hash algorithms and deprecate md5?

File: CHANGELOG.md
Comment (lpil) [inline CHANGELOG.md:38]: Let's show it being used to construct a value rather than pattern matching

File: compiler-core/src/analyse.rs
Comment (lpil) [inline compiler-core/src/analyse.rs:841]: Let's point to the constructor rather than the attribute, and then remove this field

File: compiler-core/src/error.rs
Comment (lpil) [inline compiler-core/src/error.rs:3097]: 2 space indentation please

File: compiler-core/src/parse.rs
Comment (lpil) [inline compiler-core/src/parse.rs:2143]: Always pattern match rather than checking a bool expression and then checking it a second time with `expect`

File: compiler-core/src/type_/tests/custom_types.rs
Comment (lpil) [inline compiler-core/src/type_/tests/custom_types.rs:72]: let's make it an error to deprecate a variant on a type that has been entirely deprecated

**Review by lpil — COMMENTED** [review]
Very nice!! I've left some small comments

**Comment by Wilbert-mad** [comment]
Finished!

File: compiler-core/src/analyse.rs
Comment (lpil) [inline compiler-core/src/analyse.rs:889]:
```suggestion
        // Check if all constructors are deprecated if so error.
```

File: compiler-core/src/analyse.rs
Comment (lpil) [inline compiler-core/src/analyse.rs:899]:
```suggestion
        // If any constructor record/varient is deprecated while
```

File: compiler-core/src/analyse.rs
Comment (lpil) [inline compiler-core/src/analyse.rs:907]:
```suggestion
            // Report error on all variants attibuted with deprecated
```

File: compiler-core/src/error.rs
Comment (lpil) [inline compiler-core/src/error.rs:3142]:
```suggestion
                                title: "All variants of custom type deprecated.".into(),
```

File: compiler-core/src/error.rs
Comment (lpil) [inline compiler-core/src/error.rs:3162]:
```suggestion
                                title: "Custom type already deprecated".into(),
```

File: compiler-core/src/error.rs
Comment (lpil) [inline compiler-core/src/error.rs:3159]:
```suggestion
                            let text = wrap("This custom type has already been deprecated, so deprecating \
one of its variants does nothing.

Consider removing the deprecation attribute on the variant.");
```

File: compiler-core/src/parse/error.rs
Comment (lpil) [inline compiler-core/src/parse/error.rs:405]:
```suggestion
    UnknownAttributeRecordVariant // an attribute was used that is not know for a custom type variant
```

File: compiler-core/src/type_/error.rs
Comment (lpil) [inline compiler-core/src/type_/error.rs:883]:
```suggestion
    VariantWithDeprecatedAnnotation,
```

File: compiler-core/src/warning.rs
Comment (lpil) [inline compiler-core/src/warning.rs:1037]:
```suggestion
                            "Deprecating individual custom type variants was"
```

File: compiler-core/src/type_/error.rs
Comment (lpil) [inline compiler-core/src/type_/error.rs:998]:
```suggestion
            | Error::AllVariantsDeprecated { location }
```

File: compiler-core/src/type_/error.rs
Comment (lpil) [inline compiler-core/src/type_/error.rs:999]:
```suggestion
            | Error::DeprecatedVariantOnDeprecatedType { location } => location.start,
```

**Review by lpil — COMMENTED** [review]
Fab! I've left some notes inline improving the wording of messages and making internal types more accurate. After that we're ready to merge! Thank you

**Comment by lpil** [comment]
Hi @Wilbert-mad, you've force pushed to an older version of the code so there's now merge conflicts. Please rebase on main to return to the latest version. Thank you

File: compiler-core/src/type_/error.rs
Comment (giacomocavalieri) [inline compiler-core/src/type_/error.rs:898]: Since this will appear in 1.7, could you update `VariantWithDeprecatedAnnotation` to require version `Version::new(1, 7, 0)`?

**Review by lpil — APPROVED** [review]
Thank you!!

**Pull request merged**
